tests/test_connection.py

```diff
@@ -0,0 +1,117 @@
+import pathlib
+
+import pytest
+
+from compredict.connection import Connection
+from compredict.exceptions import ClientError, ServerError
+
+
+def test_set_token(connection):
+    token = "hahsduasfbfiuyer"
+    connection.set_token(token)
+
+    expected_token = 'Token hahsduasfbfiuyer'
+
+    assert connection.headers['Authorization'] == expected_token
+
+
+def test_handle_response_with_raising_client_error(connection_with_fail_on_true, response_400):
+    with pytest.raises(ClientError):
+        connection_with_fail_on_true._Connection__handle_response(response_400)
+
+
+def test_handle_response_with_raising_server_error(connection_with_fail_on_true, response_500):
+    with pytest.raises(ServerError):
+        connection_with_fail_on_true._Connection__handle_response(response_500)
+
+
+def test_handle_response_with_last_error(connection, response_400, response_500):
+    response_400 = connection._Connection__handle_response(response_400)
+    response_500 = connection._Connection__handle_response(response_500)
+
+    assert response_400 is False
+    assert response_500 is False
+
+
+def test_handle_response_with_graph(connection, response_200_with_url, mocker):
+    mocked_wrapper = mocker.patch('tempfile._TemporaryFileWrapper')
+
+    connection._Connection__handle_response(response_200_with_url)
+
+    assert mocked_wrapper.called is True
+
+
+def test_handle_successful_response(connection, response_200):
+    actual_response = connection._Connection__handle_response(response_200)
+    expected_response = {
+        "error": "False",
+        "result": "some result"
+    }
+    assert actual_response == expected_response
+
+
+def test_successful_POST(connection, response_200, mocker):
+    endpoint = "/some/additional/endpoint"
+    data = {"data": "here we have some data"}
+    mocker.patch('requests.post', return_value=response_200)
+    actual_response = connection.POST(endpoint=endpoint, data=data)
+
+    expected_response = {
+        "error": "False",
+        "result": "some result"
+    }
+    assert actual_response == expected_response
+
+
+def test_successful_POST_with_file(connection, response_200, mocker, data):
+    file = pathlib.Path(__file__).parent.resolve().joinpath('example.json')
+    mocker.patch('requests.post', return_value=response_200)
+    content_type = "json/apllication"
+    connection.headers["Content-Type"] = content_type
+    actual_result = connection.POST(endpoint="not/as/important/endpoint/here", data=data,
+                                    files=file)
+    expected = {'error': 'False', 'result': 'some result'}
+    assert actual_result == expected
+    with pytest.raises(KeyError):
+        type = connection.headers['Content-Type']
```

Comment:
I don't understand the reason behind the with statement and the type variable, it is never used. Additionally, type is a reserved word in python, you shouldn't override it.

Reply:
Variable is not needed, that's true and also the name is bad defined. I overlooked it.  With statement is typical in pytest when you check for Exception raising. It can be done with try and except but it's somehow cleaner with pytest.raises.  It's less code.